gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -1086,15 +1008,17 @@ private int getQuotaForUser(String user) {
       ExecutionStatus jobStatus = DagManagerUtils.getExecutionStatus(dagNode);
       log.info("Job {} of Dag {} has finished with status {}", jobName, dagId, jobStatus.name());
 
-      releaseQuota(dagNode);
-
-      if (this.metricContext != null) {
-        getRunningJobsCounter(dagNode).dec();
-        getRunningJobsCounterForUser(dagNode).forEach(ContextAwareCounter::dec);
+      // Only decrement counters and quota for jobs that actually ran on the executor, not from a GaaS side failure/skip event
+      if (this.runningDags.contains(dagId)) {
+        quotaManager.releaseQuota(dagNode);
+        runningDags.remove(dagId);
```

Comment:
this check-then-set pattern doesn't look thread-safe.  is there some additional context that ensures exactly one thread would only ever work w/ the same `dagId`.  if so, please leave a comment explaining for maintainers.

Reply:
Yes only one thread would work with the same dagId, even if there are concurrent executions with the same flowname and flowgroup it will have a unique flowexecutionId that's generated from the timestamp. I'll add a comment for this